pandas/core/groupby/generic.py

```diff
@@ -1569,6 +1574,220 @@ def func(df):
+                "DataFrameGroupBy.value_counts only handles axis=0"
+            )
+
+        with self._group_selection_context():
+            df = self.obj
+
+            # Check for index rather than column grouping
+            index_grouping = self.grouper.groupings[0].name is None
+
+            # Try to find column names
+            if index_grouping:
+                keys: Any = []
+                remaining_columns = self._selected_obj.columns
+            elif isinstance(self._selected_obj, Series):
+                keys = [grouping.name for grouping in self.grouper.groupings]
+                remaining_columns = [self._selected_obj.name]
+            else:
+                if isinstance(self.keys, ops.BaseGrouper):
+                    keys = [grouping.name for grouping in self.keys.groupings]
+                elif isinstance(self.keys, str):
+                    keys = [self.keys]
+                else:
+                    keys = self.keys
+
+                remaining_columns = [
+                    key for key in self._selected_obj.columns if key not in keys
+                ]
+
+            if subset is not None:
+                remaining_columns = [key for key in subset if key not in keys]
+
+            if dropna:
+                df = df.dropna(subset=remaining_columns, axis="index", how="any")
+
+            # Add the remaining_column keys to the main grouper
+            main_grouper, _, _ = get_grouper(
+                df,
+                key=self.keys,
+                axis=self.axis,
+                level=self.level,
+                sort=self.sort,
+                mutated=self.mutated,
+                dropna=dropna,
+            )
+            groupings = list(main_grouper.groupings)
+            for key in remaining_columns:
+                grouper, _, _ = get_grouper(
+                    df,
+                    key=key,
+                    axis=self.axis,
+                    level=self.level,
+                    sort=self.sort,
+                    mutated=self.mutated,
+                    dropna=dropna,
+                )
+                groupings += list(grouper.groupings)
+
+            # Take the size of the overall columns
+            result = df.groupby(
+                groupings,
+                as_index=self.as_index,
+                sort=self.sort,
+                observed=self.observed,
+                dropna=dropna,
+            ).size()
+
+            # Change the nameof the size result
+            if self.as_index:
+                result.name = RESULT_NAME
+            else:
+                result.rename({"size": RESULT_NAME}, axis=1, inplace=True)
+
+            if normalize:
+                # Normalize the results by dividing by the original group sizes
+                indexed_group_size = df.groupby(
+                    main_grouper, sort=self.sort, observed=self.observed, dropna=dropna
+                ).size()
+                if self.as_index:
```

Comment:
Can we pass `as_index=True` above always, use this block to normalize, and then call reset_index when `as_index` is False?

Reply:
I am not quite sure what you mean but I don't think so. The as_index value results in a completely different data type and structure.
With the default as_index=True then size() and value_counts() return a Series.
With as_index=False then the result is a DataFrame with the same columns as the original plus a new one with the results, that I have labeled "count".
So both normalize and sort have to be done completely differently in the two cases.

Reply:
When is it the case that 

```
result = df.groupby(keys, as_index=False).value_counts()
expected = df.groupby(keys).value_counts().reset_index()
pd.testing.assert_frame_equal(result, expected)
```

should raise? I'm seeing this PR does raise due to the order of the results, but I don't believe that is intentional.

Reply:
I wouldn't be surprised if they were different, since I use GroupBy.size() internally. I regarded value_counts() as a sort of extension of size(), and therfore following its behaviour in the two cases.

That being said, I have tried a variety of dfs and key combinations and can't reproduce the order change you describe. Please could you give me a complete example.

Reply:
> I wouldn't be surprised if they were different,...

Is that the behavior we want? My expectation is that `as_index` would simply change where the groups are place (index vs columns), and not the order of the result. Am I missing something?

> I regarded value_counts() as a sort of extension of size(), and therfore following its behaviour in the two cases.

I cannot replicate this behavior (value of as_index changing result order) with `.size`, and if I could, would think it is a bug.

> Please could you give me a complete example.

Sure! I've played around a bunch, and `size, bins = 17, 3` in the code below seems to be minimal.

```
pd.DataFrame(
    {
        "a": list("10110210212102210"),
        "b": list("02112210220221000"),
        "c": list("01120201020021212"),
    }
).astype(int)
```
Produced by

<details>

```
import numpy as np
import pandas as pd
import pandas._testing as tm

size, bins = 17, 3
for _ in range(1000):
    df = pd.DataFrame({k: np.random.randint(bins, size=size) for k in 'abc'})
    result = df.groupby(['a'], as_index=False).value_counts()
    expected = df.groupby(['a']).value_counts().reset_index()
    # result = result.sort_values(list('abc')).reset_index(drop=True)
    # expected = expected.sort_values(list('abc')).reset_index(drop=True)
    try:
        tm.assert_frame_equal(result, expected)
    except Exception:
        print(pd.concat([result, expected], axis=1))
        a = ''.join(df.a.astype(str).tolist())
        b = ''.join(df.b.astype(str).tolist())
        c = ''.join(df.c.astype(str).tolist())
        rep = (
            f'pd.DataFrame(\n'
            '    {\n'
            f'        "a": list("{a}"),\n'
            f'        "b": list("{b}"),\n'
            f'        "c": list("{c}"),\n'
            '    }\n'
            ').astype(int)\n'
        )
        tm.assert_frame_equal(eval(rep), df)
        print(rep)

        break

```

</details>

Reply:
Wanted to add that if we do agree order of results should be the same for either value of `as_index`, then I think we can significantly simplify the implementation by following what `_wrap_aggregated_output` does, computing the result with `as_index=True` and then adjusting the index as necessary afterwards.

https://github.com/pandas-dev/pandas/blob/52c9181de6c23fa2e0143937f282fe2fde0fc6b3/pandas/core/groupby/groupby.py#L1135-L1140

Reply:
You convinced me! We are throwing away any initial index anyway.
I got rid of all the tricky frame code, work it all out with `as_index=True` and then `reset_index()` at the end. 
At least it passes all of my tests...